Now we implement the Principal component analysis from the cleaned data that we prepared in the previous notebook

In [1]:
#These are all the required libraries for the PCA.

import pandas as pd
import numpy as np

from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler


In [ ]:
df=pd.read_csv('../data/processed/cleaned_yield_changes.csv')
df.head()

,Date,6M,1Y,2Y,3Y,5Y,7Y,10Y
0,1990-01-03,0.05,0.04,0.07,0.06,0.05,0.06,0.05
1,1990-01-04,-0.04,-0.03,-0.02,-0.03,-0.01,-0.02,-0.01
2,1990-01-05,-0.05,-0.03,-0.02,0.01,0.01,0.01,0.01
3,1990-01-08,0.03,0.02,0.00,0.01,0.00,0.02,0.03
4,1990-01-09,-0.06,-0.03,0.01,-0.01,0.00,0.00,0.00


In [8]:
#We will have to drop the date column as it is not a numerical value and will not be used in the PCA. We will also have to standardize the data before applying PCA.

x=df.drop(columns=['Date'])

scaler=StandardScaler()

x_c= scaler.fit_transform(x)

#We are following mathematical notation where x is the yield changes data, x_c is the standardized yield changes data.



In [9]:
#Now we run PCA on the standardized data.

pca=PCA()
pca_results=pca.fit_transform(x_c)

In [14]:
# Now we see soem key stats

explained_variance = pd.DataFrame({
    "PC": [f"PC{i+1}" for i in range(len(pca.explained_variance_ratio_))],
    "Explained Variance": pca.explained_variance_ratio_,
    "Cumulative Variance": pca.explained_variance_ratio_.cumsum()
})

print(explained_variance)

    PC  Explained Variance  Cumulative Variance
0  PC1            0.813862             0.813862
1  PC2            0.125747             0.939609
2  PC3            0.033327             0.972935
3  PC4            0.014884             0.987820
4  PC5            0.005736             0.993555
5  PC6            0.003849             0.997404
6  PC7            0.002596             1.000000


So 3 PC's explain 97.2935% of the variance. Taking 3 PC's is a standard procedure. They denote level, slope and curvature movement respectively